src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+  }
+
+  UMPIRE_UNPOISON_MEMORY_REGION(m_allocator, ret, bytes);
+  return ret;
+}
+
+void ResourceAwarePool::deallocate(void* ptr, std::size_t size)
+{
+  auto r = getResource(ptr);
+
+  UMPIRE_LOG(Warning, fmt::format("The ResourceAwarePool requires a Camp resource. Calling deallocate with: {}.",
+                                  camp::resources::to_string(r)));
+
+  deallocate_resource(r, ptr, size);
+}
+
+void ResourceAwarePool::do_deallocate(Chunk* chunk, void* ptr) noexcept
+{
+  UMPIRE_POISON_MEMORY_REGION(m_allocator, ptr, chunk->size);
+  UMPIRE_USE_VAR(ptr);
+  chunk->free = true;
+
+  // Removing chunk from pending
+  for (auto it = m_pending_list.begin(); it != m_pending_list.end();) {
+    auto my_chunk = (*it);
+    if (my_chunk == chunk) {
+      it = m_pending_list.erase(it);
+    } else {
+      it++;
+    }
+  }
+
+  UMPIRE_LOG(Debug, "In the do_deallocate function. Deallocating data held by " << chunk);
+
+  if (chunk->prev && chunk->prev->free == true) {
+    auto prev = chunk->prev;
+
+    UMPIRE_LOG(Debug, "Removing chunk" << prev << " from free map");
+    m_free_map.erase(prev->size_map_it);
+
+    prev->size += chunk->size;
+    prev->next = chunk->next;
+
+    prev->m_event = chunk->m_event;
+    prev->m_resource = chunk->m_resource;
+
+    if (prev->next)
+      prev->next->prev = prev;
+
+    UMPIRE_LOG(Debug, "Merging with prev" << prev << " and " << chunk);
+    UMPIRE_LOG(Debug, "New size: " << prev->size);
+
+    chunk->~Chunk(); // manually call destructor
+    m_chunk_pool.deallocate(chunk);
+    chunk = prev;
+  }
+
+  if (chunk->next && chunk->next->free == true) {
+    auto next = chunk->next;
+
+    UMPIRE_LOG(Debug, "Removing chunk" << next << " from free map");
+    m_free_map.erase(next->size_map_it);
+
+    chunk->size += next->size;
+    chunk->next = next->next;
+
+    chunk->m_event = next->m_event;
+    chunk->m_resource = next->m_resource;
+
+    if (chunk->next)
+      chunk->next->prev = chunk;
+
+    UMPIRE_LOG(Debug, "Merging with next" << chunk << " and " << next);
+    UMPIRE_LOG(Debug, "New size: " << chunk->size);
+
+    next->~Chunk(); // manually call destructor
+    m_chunk_pool.deallocate(next);
+  }
+
+  UMPIRE_LOG(Debug, "Inserting chunk " << chunk << " with size " << chunk->size);
+
+  if (chunk->size == chunk->chunk_size) {
+    m_releasable_blocks++;
+    m_releasable_bytes += chunk->chunk_size;
+  }
+
+  chunk->size_map_it = m_free_map.insert(std::make_pair(chunk->size, chunk));
+}
+
+void ResourceAwarePool::deallocate_resource(camp::resources::Resource r, void* ptr, std::size_t UMPIRE_UNUSED_ARG(size))
+{
+  UMPIRE_LOG(Debug, "(ptr=" << ptr << ")");
+  UMPIRE_LOG(Debug, "(Resource=" << camp::resources::to_string(r) << ")");
+  auto chunk = (*m_used_map.find(ptr)).second;
+
+  if (chunk == nullptr) {
+    UMPIRE_ERROR(runtime_error, fmt::format("The chunk can't be found! Called deallocate with ptr: {}", ptr));
+  }
+
+  auto my_r = getResource(ptr);
```

Comment:
You already have the chunk, can't you just use chunk->resource?

Reply:
Done